### DISCUSSION TITLE
Quite literally the most complicated piece of software for no apparent reason.

### DISCUSSION THREAD

**Discussion by karnull** [wixtoolset/issues#8700] [Questions] [answered]
It's honestly quite frustrating how poorly designed this program appears to be. surely using C# on visual studio isn't the only way to release software on windows, right? apparently sodding no. sure released this oPeN sOuRcE utility for all to use, but goddamn is the documentation not just a link loop of "oh migrate to from v4 oh migrate from v3 oh download our visual studio plugin for which c# project steps are here" and what not. cant find a single page where the schematics of `Product.wxs` are discussed in a straight forward way.

no.

$ wix --version
5.0.1xxx

$ wix build Product.wxs -o installer.msi
'http://wixtoolset.org/schemas/v5/wxs'. Please make the Wix element look like the following: <Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">.

okay replaced it with v4, nope.
C:\path\to\project\Product.wxs(3) : error WIX0005: The Wix element contains an unexpected child element 'Product'.

like hell I went to 'http://wixtoolset.org/schemas/v5/wxs' to see if that'd contain any "template", but no that redirects to some 70s pop-punk rock on youtube??? how is that even legit.

wrote a program. successfully compiled. assembled binary and dlls into one location, yet cant get it to produce an installer package. like hell. swear if it wasn't a uni task I would have given up hours ago instead of sitting on my computer with 100 tabs opened (90 of which are of no use because of poor backward compatibility), remaining 10 are just hyperlinks to one other. jesus.

**Comment by OlegSirenko** [comment, marked as answer]
It looks like you don't understand the basic principles of Wix Toolset development.
**First, you need to decide which version of Wix you will use.**
From the point of view of the documentation, Wix v3 is a much more preferable option. You can find the documentation at this link: https://www.firegiant.com/docs/wix/v3/tutorial/. Or find the books "Wix Toolset Developer Guide" and "Wix Toolset Cookbook". 

If you want to use the 5th version, then you can find a playlist on YouTube: [dojo development Wix Toolset ](https://www.youtube.com/playlist?list=PLDlzbQXIs18slmqmdlS10_de_Cps-QRg6)

**Reply by OlegSirenko** [reply]
Also on [FireGiant](https://www.firegiant.com/docs/wix/tutorial/) you could find basic guides for Wix v4 and v5.

**Reply by jwdomes** [reply]
Then how do I get `wix build` to treat my `.wxs` file like a V5?

**Comment by jwdomes** [comment]
I have to say, I agree with @karshPrime. I have a simple question - how do I setup my .wxs file to use the v5 schema? I cannot believe that this isn't a trivial find.

`<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">` works for v4. 

How do I get v5 functionality in order to e.g. move away from GUID for `UpgradeCode`?

**Reply by robmen** [reply]
WiX v4, v5, v6 all use the same namespace because feedback was that changing it is too much of a breaking change.

You use v5 by using the v5 tools. Like:

```xml
<Project Sdk="WixToolset.Sdk/5.0.2">
</Project>
```

Have you read through the [Quick Start](https://docs.firegiant.com/quick-start/) and into the [Tutorial](https://docs.firegiant.com/wix/tutorial/)?

**Reply by robmen** [reply]
Note: I think when @karshPrime wrote his diatribe, that was before we had the [Quick Start](https://docs.firegiant.com/quick-start/). We put that up to make it easy to get started with the latest version... which is what we recommend if you're just getting started.

**Reply by jwdomes** [reply]
I appreciate the response. However, the Quick Start doesn't give very much and doesn't talk about the wix CLI tool at all. The tutorial assumes VS, which I definitely don't want to use in my pipeline.

If I want to use v6 with the `wix` CLI tool, do I just need to ensure I have done `dotnet tool install --global wix --version 6.0.0` and then run `wix build`?

**Reply by jwdomes** [reply]
Also, it's confusing to say that "This year, we got rid of the last “mandatory” GUID: The upgrade code.", but then I see that:

```
ProductCode


Optional ProductCode GUID for the package. By default, a new ProductCode property will be generated with every build enabling major upgrades.
```
Is optional, in the sense that it will generate a GUID automatically, but if you try a human-readable string in `ProductCode` you get an error.

**Reply by robmen** [reply]
> assumes VS, which I definitely don't want to use in my pipeline.

Yeah, you definitely do not want to use VS in your pipeline. That's why the Quick Start shows how you build from the command-line using `dotnet build` (or you can use `msbuild -Restore`). The doc doesn't use `wix.exe` because we don't recommend using it.

> If I want to use v6 with the wix CLI tool, do I just need to ensure I have done dotnet tool install --global wix --version 6.0.0 and then run wix build?

Sure. but it is better to use a .wixproj instead, like the Quick Start shows.

> but if you try a human-readable string in ProductCode you get an error.

Yeah, you get an error because `ProductCode` must be a GUID. But it's optional (just like the doc says it is) and thus not recommended to be set. That's also why the text you quoted from the What's New page says `...rid of the last “mandatory” GUID...`. The word `"mandatory"` is important in that sentence. You can still use GUIDs if you _need_ to (usually for backwards compatibility), but we don't recommend it.

It appears you came at WiX with a bunch of assumptions about how things work and then ran into problems when you found your assumptions were incorrect. At this point, I recommend slowing down and reading more. If you have recommendations how we can improve the doc (there is plenty of work for us to do), that's great. But don't just say the doc is terrible because it didn't tell you how to do something the way you wanted to do it when that way isn't what we'd recommend anyone start with.

**Reply by jwdomes** [reply]
I definitely have preconceived notions about how I want WiX to work. My hope was that WiX would allow me to not have to think too much about the underlying MSI specifics. In my mind, I should e.g. be able to make an MSI that installs an app for all users to a particular folder with the standard shortcuts (Desktop, Start Menu) and Registry entries without having to get too "in the weeds". Unfortunately, even for what seems to me a pretty cookie-cutter need, I'm still not able to get it to work with the SDK approach. Only the `wix build` has worked thus far.

With the SDK I get errors like:

`error WIX0204: ICE43: Component MainExecutableComponent has non-advertised shortcuts. It should use a registry key under HKCU as its KeyPath, not a file.`
`error WIX0204: ICE57: Component 'MainExecutableComponent' has both per-user and per-machine data with a per-machine KeyPath.`

I'm sure all of these issues have answers/solutions. Perhaps the above errors are obvious to someone experienced in the specifics of Windows installation particulars and MSIs, but for someone who doesn't have much experience in these areas, finding those answers has been challenging.

**Reply by robmen** [reply]
Ahh, this is another example why you want to be using the SDK. Those errors appear because static validation is part of the standard build process. Using `wix.exe` you need to do both the `build` command and the `msi validate` command to get the same behavior. If you only use `wix build` then you could be shipping your MSI with static validation errors that will impact your install later.

**Reply by jwdomes** [reply]
There's a lot of chatter all over the forums about ICE57. Again, though, I don't see any straightforward explanation of how to solve this or suppress the warning. This is the kind of thing I keep running into. Seemingly endless caveats.

**Reply by robmen** [reply]
You don't provide enough context (.wxs code, in this case) for anyone here to help you.

**Reply by jwdomes** [reply]
I was eventually able to fix the issue through trial and error. I appreciate your time and responses.

**Reply by robmen** [reply]
So, after kicking off this thread by complaining about the documentation, your final contribution is "I eventually figured it out." That doesn't help people going forward. Why not provide a constructive write-up that explains the path you took so that others can learn from your experience, and more experienced developers can see the missteps beginners take, so we can see if there are opportunities to improve the documentation?

**Reply by npr-gh** [reply]
All right, I'll bite and spend a couple hours exactly showing what's wrong here. Let's say I followed the advice and tried to use WiX v3, open source, no support, for an actual project, with the standard Visual Studio addon. 

Let's assume I'm trying to do the following, which I'll consider a very minimal example, something that should take me 15 minutes if the software and plugins are any good at all. 15 days of hell later, I'm still nowhere near done with step one, getting it to output an installer at all, so your stuff is just awful, and spending a few thousand on InstallShield or what have you commercial setup-making software would've been a way better idea. 

With the old Installer for Visual studio projects, I got several of these steps _for free_, just for creating the project! It would even always automatically include all the stuff in the project output, and be connected to the VS form builder so you could just edit the installer's forms, with a License, about, Folder select, and shortcut options all there by default.

So, reducing it down to an example. I'm ACME, and I'm trying to make an .msi file to install my application 'sprocket' on the windows computer of my client Wile. E. Coyote, inside (by default, can be modified) the very typical folder 'C:\Program Files\acme\sprocket\ '. It needs to add it to the desktop via a shortcut, and the start menu (both can be optionally turned off as per usual). Its hundreds of various files and libraries are already present in the 'Release' folder of the solution on the development machine inside the project folder `C:\Projects\coyote_sprocket\bin\Release\*.*`, and work from there. The program uses a T-SQL database and needs to have Wile insert the name of the server and database instance name it's suppose to use. If Wile has multiple computers, the same should be able to be done via a CLI parameter. 

I should be able to follow a tutorial and get through most of it. I just can't. 

Using the standard Visual studio plugin, Wix 3.14, heat. I got a bunch of outputs. I got so far, by randomly bashing together command lines, to get Heat to output something resembling a wxs file. I have no idea what I'm doing, the documentation is garbage. 

Here's how I would expect this to work in the modern world: 

1. Type the company name in a box. 
2. Type the software name in a box, automatically populated from the project name. 
3. The VS plugin figures out how to wrangle it into the xml so that it can be wrangled into the MSI. 
4. Figure out where to click to add an input form, then a single textbox, dbServer. 
5. Figure out how to 'connect' the input box to a variable var.dbServer. 
6. Finally tell it to output that variable into the settings file using a simple name, 'SettingName = Server', where it puts it into 

```
<applicationSettings>
    <Sprocket.My.MySettings>
        <setting name="Server" serializeAs="String">
        </setting>
    </Sprocket.My.MySettings>
<applicationSettings>
```

7. Add a command line '-s' parameter into a table and forward it to var.dbServer too.
8. The other requirements are so bog-standard they're already configured by default. 

I've dealt with over a dozen LGHT, HEAT, and INST errors already and it just endlessly keeps spitting out more the more and more I try to rework it. Honestly this is seriously bad. 

Step #0. Spit out useful errors. Your tools don't do this. It doesn't suggest where to put things or how to do things, far from it you get arcane error codes from MSI, which isn't even the same tool, so you're left divining which Xml element matches which error in what way. In other words, you end up just trying random things until you get a different error. I mean I could learn microsoft MSI and build it by hand... but that defeats the entire purpose of WiX and having all that nicely in source control (git).

From the start, you get a template that doesn't do anything when you first let the extension generate installer code. The internet is anything but helpful (as are chatbots) due to your many conflicting versions and syntaxes and poor tutorials. Hooking up heat and so on has to be done by hand. But calling a compiler from the command line within Visual Studio, what with half a dozen different escaping and cli conventions is already hellish. Let alone three (four?) different tools all working together, all with their own error codes, none of which are properly connected.

Step #1. Clarify what are variables, what are magic words, what are examples, what are hardcoded values. Where do you set things? This is seriously lacking and my #1 source of confusion in the syntax. Walk the user through where you define the paths. And clarify which ones are which! SOURCE vs DESTINATION. Are we talking a path on the development machine or on the install target? 

Okay, so here's a magic incantation, I mean a heat command line; 

```
 "$(WIX)bin\heat.exe" dir "C:\Projects\coyote_sprocket\bin\Release\net48" -var "var.sprocket.TargetDir" -cg wkoAuto -ke -srd -dr InstallFolder -sfrag -gg -o "C:\Projects\coyote_sprocket\Installer2\wkoInstaller2\Dynamic.wxs"
```

And here's a little fragment from product.wxs that it complains about that contains sort of the 'root' (I believe???) of how to build up a web of references using Directory and DirectoryRef. I first got to point it to C:\Program Files, and the place where the Start Menu and so on are.

```
<Directory Id="TARGETDIR" Name="SourceDir">
  <Directory Id="ProgramFilesFolder">
    <Directory Id="MyCompanyDir" Name="acme">
      <Directory Id="InstallFolder" Name="sprocket"/>			
    </Directory>
  </Directory>
  <Directory Id="ProgramMenuFolder">
    <Directory Id="MenuMyCompanyDir" Name="acme">
      <Directory Id="StartMenuFolder" Name="sprocket"/>
    </Directory>
  <Directory Id="DesktopFolder" Name="Desktop" />
</Directory>
```
And an example I found for doing a setting; 

<DirectoryRef Id="InstallFolder">
	<Component Id="sprocketComponent" Guid="3f007a54-c86f-4acb-b2fa-b7862eb2297c">
		
							
		<util:XmlFile 
			Id="UpdateSqlPath" 
			Action="setValue"
			File="[InstallFolder]sprocket.exe.config"
			SelectionLanguage="XPath"
			Permanent="yes"
			ElementPath="/configuration/applicationSettings/sprocket.My.MySettings/setting[\[]@name='Server'[\] ]/@value"
			Value="[SQLPATH]"/>
    </Component>
</DirectoryRef>


It ends up building a comparatively giant 150KB XML file with a bunch of stuff in it, before then spitting out afterwards, first some warnings, 

```
1>heat.exe(0,0): warning HEAT5151: Could not harvest data from a file that was expected to be an assembly: C:\Projects\coyote_sprocket\bin\Release\net48\Azure.Core.dll. If this file is not an assembly you can ignore this warning. Otherwise, this error detail may be helpful to diagnose the failure: Could not load file or assembly 'Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
```

That doesn't bode quite well, though I've checked the `dynamic.wxs` I'm making it generate and the file references are there (at least, I can find the text 'Azure.Core.dll' inside of it).

and then some errors (I've seen and fixed many, many more of these, all from various "tutorials" online that got things wrong either because Wix changed or otherwise, I'm leaving this one in as a demonstration).

```
error LGHT0204: ICE03: Invalid format string; Table: XmlFile, Column: ElementPath, Key(s): UpdateSqlPath
 error LGHT0204: ICE18: KeyPath for Component: 'sprocketComponent' is Directory: 'InstallFolder'. The Directory/Component pair must be listed in the CreateFolders table.
```
That first error is apparently because the one page I found online with suggestions of how to implement whatever strange way of escaping WiX uses for XPath inside of XML-with-WiX-Macros was wrong.  

And here's the crux: I don't know anything about any CreateFolders table. There's no such thing here. That exists in the MSI **output**. I don't want to have to know all the little details about how Microsoft's horrible installer format works. Taming that monster is your job. So instead, I'd like to know what's wrong with my XML, and how it'll nicely put in CreateFolders entries (or whatever they're named) for all these various folders that need it (with the obvious actions of 'create iff they do not already exist' upon install or update).  

None of the tutorials properly explain where those 'root'-type directory references come from. Only how Wix itself works... but without a starting point, you can't get anywhere! You can make endless chains of directoryRefs, but in the end it can never find anything if you don't give it a starting point. Where does 'Program files' come from? 

So, I try stuff out, and eventually find that I had to delete a space in ElementPath and it should be  

```
    ElementPath="/configuration/applicationSettings/sprocket.My.MySettings/setting[\[]@name='Server'[\]]/@value"
```
And that adding this in the component makes the ICE18 go away (no idea if it's supposed to be there or why). 

```
    <CreateFolder Directory="InstallFolder" />
```

Now it spits out a setup that crashes with the mysterious 'error 2363.' And I'm left in the middle of nowhere again. 

=====

Another way to express my beef with it is that there's no minimal *working* examples out there. You need to do a tonne of very obfuscated stuff to even get it to compile, and I dread the amount of effort it'll take to make this pile of junk work. 

That's my problem.

Also see https://news.ycombinator.com/item?id=39949010. It's not just my problem.

**Reply by robmen** [reply]
I'm very curious why you started with WiX v3.14?

Also, did you not find the [Quick Start](https://docs.firegiant.com/quick-start/) on the docs site, which continues on with a more complete tutorial?

Your comments about `heat` are valid. We deprecated it a couple of versions ago and replaced it wth the much easier to use `Files` element.

The ICE messages are more challenging. Those are the standard static validation checks from the Windows Installer team. That's why the messages are much more MSI focused. We have a dream of re-writing all of them in WiX itself to provide better messages (and run much, much faster), but there are a lot of checks and they aren't public so we'd be re-implementing them by observation. Like I said, it's challenging.

Also, the WiX Toolset is designed to provide complete control over the Windows Installer. That gives you access to a lot of power but sometimes it exposes some sharper edges. Always opportunities to improve.

I'm still very curious why you'd start with WiX v3.14?
